api/envoy/service/discovery/v3/discovery.proto

```diff
@@ -288,4 +288,10 @@ message Resource {
   // testing where the fault injection should be terminated in the event that Envoy loses contact
   // with the management server.
   google.protobuf.Duration ttl = 6;
+
+  // If true, xDS proxies may not cache this resource.
+  // Note that this does not apply to non-proxy xDS clients, which must cache resources for their
+  // own use, regardless of the value of this field.
```

Comment:
Can you describe a bit more of what the behavior would be if this is set in for example xds-relay? I'm a little confused about the use case here and am wondering if we should also update the protocol docs? cc @jessicayuen @jyotima

Reply:
The use-case here is where the content is dynamically generated by the authoritative server based on some information that is out-of-band with respect to the resource name (e.g., based on the identity of the xDS client).  This is basically analogous to the way that an HTTP server can tell clients not to cache data that is dynamically generated.

In this case, a caching proxy like xds-relay should not be allowed to cache the resource, because the resource name alone would not be sufficient as a unique cache key.  Different clients requesting the same resource name should actually get different contents, so the only way clients can get the right thing is to get it directly from the authoritative server.

I would expect this to be used only in cases where a caching proxy is not expected to be used.  The actual purpose of this field is to basically prevent from accidentally caching something that should not be cached (e.g., if someone forgets that the resources are dynamically generated and tries to add a caching proxy).  I think it would be fine for xds-relay to ignore any such resource that it might get from the authoritative server, and xds-relay can respond to its clients as if the resource does not exist.

I don't expect that this is something that would be used very often, and I would actually prefer not to add it at all.  But unfortunately, we have a use-case where we need to do something like this, and it's a lot easier to add this escape hatch than it is to rearchitect such that different clients use different resource names.

Reply:
OK this makes sense to me. @jessicayuen @jyotima any concerns?

Reply:
The use case makes sense to me, but I think the description here is confusing. "non-proxy xDS clients" is particularly confusing because it sounds like only applicable to gRPC and Envoy Mobile, but Envoy is also falls into this category.

Also, should we name this field `dynamically_generated`? IMO that reflect what it indicates.

Reply:
Good point, @lizan.  I've clarified the wording here to specify that this is talking about xDS proxies, not xDS clients that happen to be data plane proxies. :)

I kind of prefer the name `do_not_cache`, because I think that's a more obvious indication of what should be done with the resource.  It may not be obvious to someone seeing `dynamically_generated` that that means that it can't be cached.